storage/testonly/admin_storage_tester.go

```diff
@@ -126,25 +127,199 @@ func (tester *AdminStorageTester) TestCreateTree(t *testing.T) {
 		}
 
 		// Make sure a tree was correctly stored
-		ctx = context.Background()
-		readOnlyTX, err := s.Snapshot(ctx)
+		storedTree, err := getTree(s, newTree.TreeId)
 		if err != nil {
-			t.Fatalf("%v: Snapshot() = %v, want = nil", i, err)
-		}
-		defer readOnlyTX.Close()
-		storedTree, err := readOnlyTX.GetTree(ctx, newTree.TreeId)
-		if err != nil {
-			t.Errorf(":%v: GetTree() = (%v, %v), want = (%v, nil)", i, newTree.TreeId, err, newTree)
+			t.Errorf(":%v: getTree() = (%v, %v), want = (%v, nil)", i, newTree.TreeId, err, newTree)
 			continue
 		}
 		wantTree = *storedTree
 		if !reflect.DeepEqual(newTree, &wantTree) {
 			t.Errorf("%v: newTree = \n%v, wantTree = \n%v", i, newTree, &wantTree)
 		}
-		if err := readOnlyTX.Commit(); err != nil {
-			t.Errorf("%v: Commit() = %v, want = nil", i, err)
+	}
+}
+
+// TestUpdateTree tests AdminStorage Tree updates.
+func (tester *AdminStorageTester) TestUpdateTree(t *testing.T) {
+	s := tester.NewAdminStorage()
+
+	unrelatedTree, err := createTree(s, MapTree)
+	if err != nil {
+		t.Fatalf("createTree() = (_, %v), want = (_, nil)", err)
+	}
+
+	referenceLog := *LogTree
+
+	validLog := referenceLog
+	validLog.TreeState = trillian.TreeState_FROZEN
+	validLog.DisplayName = "Frozen Tree"
+	validLog.Description = "A Frozen Tree"
+	validLogFunc := func(t *trillian.Tree) {
+		t.TreeState = validLog.TreeState
+		t.DisplayName = validLog.DisplayName
+		t.Description = validLog.Description
+	}
+
+	invalidLogFunc := func(t *trillian.Tree) {
+		t.TreeState = trillian.TreeState_UNKNOWN_TREE_STATE
+	}
+
+	readonlyChangedFunc := func(t *trillian.Tree) {
+		t.TreeType = trillian.TreeType_MAP
+	}
+
+	referenceMap := *MapTree
+	validMap := referenceMap
+	validMap.DisplayName = "Updated Map"
+	validMapFunc := func(t *trillian.Tree) {
+		t.DisplayName = validMap.DisplayName
+	}
+
+	// Test for an unknown tree outside the loop: it makes the test logic simpler
+	if _, errOnUpdate, err := updateTree(s, -1, func(t *trillian.Tree) {}); err == nil || !errOnUpdate {
+		t.Errorf("updateTree() = (_, %v, %v), wanted lookup error (ie, error on update)", errOnUpdate, err)
+	}
+
+	tests := []struct {
+		desc                 string
+		createTree, wantTree *trillian.Tree
+		updateFunc           func(*trillian.Tree)
+		wantErr              bool
+	}{
+		{
+			desc:       "validLog",
+			createTree: &referenceLog,
+			updateFunc: validLogFunc,
```

Comment:
(maybe) given they're only used once I wonder if it might be more readable to have the function definitions in the table.